File: 0328_Odd_Even_Linked_List/try_6.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        tmp_odd = odd = ListNode()
        tmp_even = even = ListNode()
        
        a = ListNode()
        a.next = head
        
        isOdd = True
        while a and a.next:
            b = a.next
            a.next = None
            a = b
            
            
            if isOdd:
                tmp_odd.next = b
                tmp_odd = tmp_odd.next
            else:
                tmp_even.next = b
                tmp_even = tmp_even.next
            isOdd = not isOdd
            
        tmp_odd.next = even.next
        return odd.next

File: 0206_Reverse_Linked_List/try_6.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function reverseList(head: ListNode | null): ListNode | null {
    // O(n) / O(n) - where n is number of nodes in the linkedlist
    if (head === null) {
        return null;
    }
    
    let tail = null;
    function recr(node: ListNode | null) {
        if (node.next === null) {
            tail = node;
            return node;
        }
        
        recr(node.next).next = node;
        node.next = null;
        return node;
    }
    
    recr(head);
    return tail;
};
